core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -117,9 +127,24 @@
   }
 
   private final class ContextCancellationListener implements CancellationListener {
+    // This is NOT null only if the stream cancellation need to be delayed in the case of
+    // deadline-exceeded.
+    @Nullable private Listener<RespT> observer;
+
+    private void setObserver(Listener<RespT> observer) {
+      this.observer = observer;
+    }
+
     @Override
-    public void cancelled(Context context) {
-      stream.cancel(statusFromCancelled(context));
+    public void cancelled(final Context context) {
+      if (context.getDeadline() == null || !context.getDeadline().isExpired()) {
+        stream.cancel(statusFromCancelled(context));
+      } else {
+        if (observer != null) {
```

Comment:
I'm not very comfortable with this doing nothing if observer is null. It looks like an optimization, but it requires multiple pieces to behave in a particular way. I think it would be better to add the optimization to `delayedCancelOnDeadlineExceeded()`: if `deadlineCancellationSendToServerFuture` is already non-null (the method has already been called), return early. Note there's a race there, but there's a race there already (and so I'll have another comment about it elsewhere).

Changing this would also mean we could pass the `Listener` in the constructor and get rid of `setObserver()`.